graph/pull.go

```diff
@@ -641,20 +640,32 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
 		}(&downloads[i])
 	}
 
+	// Register the images serially. We must wait for all downloads to complete if one fails,
+	// in order to cleanup the others.
+	var lastErr error
 	var tagUpdated bool
 	for i := len(downloads) - 1; i >= 0; i-- {
-		d := &downloads[i]
-		if d.err != nil {
-			if err := <-d.err; err != nil {
-				return false, err
+		registerFunc := func(d *downloadInfo, pullAborted bool) error {
+			// Cleanup is guranteed to be invoked only after the download routine exits.
+			defer func() {
+				if d.tmpFile != nil {
+					d.tmpFile.Close()
+					os.Remove(d.tmpFile.Name())
+				}
+			}()
+
+			if d.err != nil {
+				if err := <-d.err; err != nil {
```

Comment:
If there was prior error, this is going to wait for each of the other operations to finish. Why not check pullAborted before this, would also ensure the same error is returned.